pathway-analysis/03-gene_set_variation_analysis.Rmd

```diff
@@ -0,0 +1,295 @@
+# number
+lengths_vector <- purrr::map(random_gene_sets, ~ length(.x)) %>%
+  purrr::map(~ paste0("pathway_", .x)) %>%
+  # Return a vector
+  purrr::flatten_chr()
+
+# Add the names in lengths_vector to the list
+random_gene_sets <- random_gene_sets %>% 
+  # make.names() appends a "version" if something is not unique
+  purrr::set_names(nm = make.names(lengths_vector, unique = TRUE))
+```
+ 
+Run GSVA on our dataset with the same parameters as before, but now with random gene sets.
+
+```{r random_gsva, live = TRUE}
+random_gsva_results <- gsva(rnaseq_mat, 
+                            random_gene_sets,
+                            method = "gsva",
+                            # Appropriate for our transformed data on 
+                            # log2-like scale
+                            kcdf = "Gaussian",  
+                            # Minimum gene set size
+                            min.sz = 15, 
+                            # Maximum gene set size
+                            max.sz = 500,
+                            # Compute Gaussian-distributed scores
+                            mx.diff = TRUE,
+                            # Use 4 cores
+                            parallel.sz = 4)
+```
+
+Now let's make a plot to look at the distribution of scores from random gene sets.
+First we need to get this data in an appropriate format for `ggplot2`.
+
+```{r longer_random_gsva}
+# The random results are a matrix
+random_long_df <- random_gsva_results %>%
+  data.frame() %>%
+  # Gene set names are rownames
+  tibble::rownames_to_column("gene_set") %>%
+  # Get into long format
+  tidyr::pivot_longer(cols = -gene_set, 
+                      names_to = "Kids_First_Biospecimen_ID",
+                      values_to = "gsva_score") %>%
+  # Remove the .version added by make.names()
+  dplyr::mutate(gene_set = stringr::str_remove(gene_set, "\\..*")) %>%
+  # Add a column that keeps track of the gene set size
+  dplyr::mutate(gene_set_size = stringr::word(gene_set, 2, sep = "_")) %>%
+  # We want to plot smallest no. genes -> largest no. genes
+  dplyr::mutate(gene_set_size = factor(gene_set_size, 
+                                       levels = c(15, 25, 50, 100, 250, 500)))
+
+```
+
+Let's make a violin plot so we can look at the distribution of scores by gene set size.
+
+```{r random_violin}
+# Violin plot comparing GSVA scores of different random gene set sizes
+random_long_df %>%
+  ggplot2::ggplot(ggplot2::aes(x = gene_set_size, 
+                               y = gsva_score)) +
+  # Make a violin plot
+  ggplot2::geom_violin() +
+  # Add a point with the mean value
+  ggplot2::stat_summary(
+    geom = "point",
+    fun.y = "mean"
+  ) +
+  # Flip the axes
+  ggplot2::coord_flip() +
+  ggplot2::labs(title = "Random Gene Set GSVA Scores",
+                x = "gene set size",
+                y = "GSVA score") +
+  ggplot2::theme_bw()
+```
+
+What do you notice about these distributions?
```

Comment:
I want to suggest something about getting to significance through randomization here. Is that beyond the scope?
Does GSVA implement something like this?

```suggestion
What do you notice about these distributions?
How might you use this information to inform your interpretation of GSVA scores?
```

Reply:
I don't think GSVA implements anything like this. The way I think about "significance" of pathway scores in this vein is testing for differential expression if you happen to have group information, rather than an individual sample's score being different from a null distribution.  Quoting from the GSVA paper:

> Although the GSVA algorithm itself does not evaluate statistical significance for the enrichment of gene sets, significance with respect to a phenotype can be easily evaluated using conventional statistical models. Likewise, false discovery rates can be estimated by permuting the sample labels (Methods). We make no general prescription for thresholds of significance or false discovery, as these choices are highly context dependent and may vary according to each experiment.

I guess my point was more that you may want to make your scores more comparable (maybe scaling would be enough, without generating random gene sets...) if you're going to try to compare between pathways. Was that clear? It is after 5pm on Friday 😅

Reply:
I can add your suggested sentence either way!